localstack/plugin/manager.py

```diff
@@ -0,0 +1,316 @@
+        return name in self._plugins
+
+    def is_loaded(self, name: str) -> bool:
+        return self._require_plugin(name).is_loaded
+
+    @property
+    def _plugins(self) -> Dict[str, PluginContainer[P]]:
+        if self._plugin_index is None:
+            with self._init_mutex:
+                if self._plugin_index is None:
+                    self._plugin_index = self._init_plugin_index()
+
+        return self._plugin_index
+
+    def _require_plugin(self, name: str) -> PluginContainer[P]:
+        if name not in self._plugins:
+            raise ValueError("no plugin named %s in namespace %s" % (name, self.namespace))
+
+        return self._plugins[name]
+
+    def _load_plugin(self, container: PluginContainer):
+        with container.lock:
+
+            # instantiate Plugin from spec if necessary
+            if not container.is_init:
+                try:
+                    LOG.debug("instantiating plugin %s", container.plugin_spec)
+                    container.plugin = container.plugin_spec.factory()
+                    container.is_init = True
+                    self._fire_on_init_after(container.plugin)
+                except Exception as e:
+                    if LOG.isEnabledFor(logging.DEBUG):
+                        LOG.exception("error instantiating plugin %s", container.plugin_spec)
+
+                    self._fire_on_init_exception(container.plugin_spec, e)
+                    container.init_error = e
+                    return
+
+            plugin = container.plugin
+
+            if not plugin.should_load():
+                LOG.debug("not loading deactivated plugin %s", container.plugin_spec)
+                return
+
+            args = self.load_args
+            kwargs = self.load_kwargs
+
+            self._fire_on_load_before(plugin, args, kwargs)
+            try:
+                LOG.debug("loading plugin %s:%s", self.namespace, container.plugin_spec.name)
+                result = plugin.load(*args, *kwargs)
+                self._fire_on_load_after(plugin, result)
+                container.load_value = result
+                container.is_loaded = True
+            except Exception as e:
+                if LOG.isEnabledFor(logging.DEBUG):
+                    LOG.exception("error loading plugin %s", container.plugin_spec)
+                self._fire_on_load_exception(plugin, e)
+                container.load_error = e
+
+    def _init_plugin_index(self) -> Dict[str, PluginContainer]:
+        return {plugin.name: plugin for plugin in self._import_plugins() if plugin}
+
+    def _import_plugins(self) -> List[PluginContainer]:
+        return [
+            self._create_container(spec)
+            for spec in self.finder.find_plugins()
+            if spec.namespace == self.namespace
+        ]
+
+    def _create_container(self, plugin_spec: PluginSpec) -> PluginContainer:
+        container = PluginContainer()
+        container.lock = threading.Lock()
```

Comment:
Couldn't fully wrap my head around the different corner cases, but would it be possible that - as part of calling the `.factory()` of one plugin - we end up calling `_load_plugin(..)` of a `container` that has already been loaded (although that would technically indicate a circular dependency (?)).

In general, wondering if we should use a reentrant lock here, to allow recursive calls from the same thread.
```
container.lock = threading.RLock()
```
(I generally tend to prefer `RLock` over `Lock`, as recursion/circular dependency errors are often easier to debug than locking issues.)

Reply:
seems like a good idea to me! :+1: not sure if there are any cases where one thread would like to release the container lock held by another thread. because that wouldn't be possible with the `RLock` (can only be released by the thread that acquired it). but i think deadlock safety is more important

Reply:
nit: I think we have the type hint as `RLock` now, but the instance still gets created as `Lock()`.